### ISSUE TITLE
[Impeller] CupertinoDatePicker might cause text on screen to glitch.

### ISSUE THREAD

**Issue by hexlogic** [flutter/flutter#144331]
### Steps to reproduce

1. Add CupertinoDatePicker wrapped in showCupertinoModalPopup
2. Try to scroll dates

### Expected results

No glitches in text

### Actual results

Text is randomly glitching

### Code sample

<details open><summary>Code sample</summary>

```dart
    await showCupertinoModalPopup(
      barrierDismissible: true,
      context: context,
      builder: (context) {
        return Container(
          height: 250,
          color: Colors.white,
          child: Column(
            children: [
              Flexible(
                child: CupertinoDatePicker(
                  mode: CupertinoDatePickerMode.date,
                  initialDateTime: DateTime.now(),
                  onDateTimeChanged: (date) {
                    setState(() {
                      _date = date;
                    });
                  },
                ),
              ),
            ],
          ),
        );
      },
    );
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/15138090/671c6558-e615-4604-a09a-d3aaab73df96

</details>

### Logs

<details open><summary>Logs</summary>

https://pastebin.com/6pHNQP0v

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.19.0, on macOS 14.3.1 23D60 darwin-arm64, locale en-UA)
    • Flutter version 3.19.0 on channel stable at /Users/admin/devkit/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision bae5e49bc2 (2 weeks ago), 2024-02-13 17:46:18 -0800
    • Engine revision 04817c99c9
    • Dart version 3.3.0
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/admin/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/admin/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C65
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.86.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] VS Code (version 1.87.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.64.0

[✓] Connected device (5 available)
    • iPhone 15 Pro (mobile) • 42E31902-8A83-4F08-8567-7878BDCA59D5 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-17-2 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 14.3.1 23D60 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 122.0.6261.69

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by sudhar08** [comment]
hey @hexlogic  on side is working fine can you  provide the video in mp4 format and screenshots of the issue your facing rightnow

**Comment by zhangtingtingx1** [comment]
Same here with stable 3.19.0
Text in the whole page might glitch like this:
![IMG_0175](https://github.com/flutter/flutter/assets/2739552/ecc32f34-e2e2-4f3a-b8d6-f06b154b70cb)

**Comment by danagbemava-nc** [comment]
Hi @hexlogic, what device(s) are you seeing this on? How often does this bug occur? I can't seem to reproduce this on my device.

**Comment by danagbemava-nc** [comment]
Hi @zhangtingtingx1, can you provide a complete minimal reproducible code sample so that we can investigate this issue?

**Comment by danagbemava-nc** [comment]
Hi @hexlogic, did you experience this prior to upgrading to stable 3.19.0? How many times have you experienced this so far? 

Does it come with extended usage of the app or it can come at any point at all?

**Comment by hexlogic** [comment]
> Hi @hexlogic, did you experience this prior to upgrading to stable 3.19.0? How many times have you experienced this so far? 
> 
> 
> 
> Does it come with extended usage of the app or it can come at any point at all? 

Yes, it's been even upgraded to 3.19.2
I've got this or similar behaviour in 4/10 cases. It can come at any point

**Comment by Nurik7** [comment]
+1 here, I'm experiencing the same issue when I use an app on debug or release versions with Impeller enabled. I think that there's some issue with Impeller itself, because I didn't experience this behavior when I ran it with --no-enable-impeller flag

Btw, can I use manual SkSL warm-up for now? Until the problem gonna be solved

**Comment by hexlogic** [comment]
> +1 here, I'm experiencing the same issue when I use an app on debug or release versions with Impeller enabled. I think that there's some issue with Impeller itself, because I didn't experience this behavior when I ran it with --no-enable-impeller flag
> 
> 
> 
> Btw, can I use manual SkSL warm-up for now? Until the problem gonna be solved

Yes, you're good to go. This is actually what we did on our end.

**Comment by Nurik7** [comment]
> > +1 here, I'm experiencing the same issue when I use an app on debug or release versions with Impeller enabled. I think that there's some issue with Impeller itself, because I didn't experience this behavior when I ran it with --no-enable-impeller flag
> > Btw, can I use manual SkSL warm-up for now? Until the problem gonna be solved
> 
> Yes, you're good to go. This is actually what we did on our end.

Sorry, but can you tell me your current flutter version? I'm trying to warmup shaders by pressing "M", but i'm getting this error:

**Comment by danagbemava-nc** [comment]
Hi all, we would appreciate it if we could get a complete minimal reproducible code sample that can consistently reproduce the issue. So far I have only been able to reproduce it once with the code sample below.

Labeling for awareness for now. 


<details>
<summary>code used</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  DateTime? _date;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Center(
          child: SingleChildScrollView(
        child: Text('''
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse vitae arcu quam. Donec ut augue viverra arcu pulvinar fringilla. Sed odio ante, semper nec egestas sed, pretium eu erat. Suspendisse nisi urna, sodales vitae efficitur non, commodo in libero. Mauris dapibus urna massa, sed finibus orci dignissim ut. Donec feugiat semper ipsum, efficitur venenatis risus egestas eget. Aliquam ac mi sed est aliquam efficitur nec quis nulla. Praesent elit risus, maximus sit amet augue at, varius semper nibh. Praesent volutpat tellus mi, id congue orci tempus eget.
      
      Curabitur sed fringilla tellus, sed tincidunt ligula. Donec at elementum justo. Morbi vitae felis scelerisque, gravida risus tincidunt, auctor elit. Curabitur ultrices tortor id euismod facilisis. Praesent aliquam, turpis at viverra ultricies, dui diam egestas urna, sit amet lacinia lorem mauris vel orci. Duis tincidunt sed lectus eu posuere. Ut viverra massa porttitor, dignissim mi vitae, pharetra sem. Fusce quis quam sit amet felis volutpat tincidunt. Phasellus pulvinar, dolor sed sodales faucibus, mi purus interdum quam, id auctor odio felis mattis lacus. Aenean elit augue, tincidunt ut tempor vel, maximus vel quam. Donec pulvinar turpis finibus mi molestie, et bibendum nulla ultricies. Aliquam erat volutpat. Morbi semper sem purus.
      
      Nullam facilisis, neque et maximus scelerisque, nulla tortor dictum nisi, vel rhoncus metus est at mi. Donec ut rhoncus libero. Nam quis justo at urna egestas cursus eu non magna. Cras semper dapibus elit sit amet imperdiet. Cras in dignissim nulla. Nullam vitae malesuada velit. Sed facilisis nunc ac urna rhoncus rhoncus.
      
      Aenean in sollicitudin mi, in vehicula enim. Nulla facilisi. Nunc scelerisque pulvinar elit ut placerat. Ut nulla massa, pellentesque a semper eu, dignissim non ligula. Nulla vel ultrices lorem. Vestibulum sit amet iaculis nisl. Nulla ut lacus luctus, dignissim velit nec, egestas lacus. Ut euismod tortor ut purus ultricies, vitae scelerisque felis bibendum. Morbi iaculis nec turpis eu tempus. Nulla vitae ipsum nec neque interdum hendrerit id vel ante. Donec mattis volutpat viverra. Etiam id convallis lacus, at faucibus dui. Praesent enim tellus, tristique a pellentesque at, maximus eu massa. Nullam mattis porta est sed finibus. Fusce at facilisis felis, et cursus turpis. Praesent cursus leo consectetur ipsum tincidunt eleifend.
      
      Fusce pharetra neque ac tellus fermentum, in semper sem venenatis. In egestas mattis varius. Pellentesque vel libero non felis euismod fermentum. Donec id nulla vel eros sodales vestibulum. Vestibulum tempor orci ac elit ultrices rhoncus. Nulla sollicitudin metus orci, ultrices sodales ex dapibus sit amet. Praesent auctor tincidunt nisl, id finibus sem iaculis in. Phasellus mattis ornare orci, eu dictum nisi feugiat quis. Morbi vel nulla a libero vulputate auctor. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Nunc nec justo imperdiet dui mollis aliquet. Ut ante turpis, elementum pharetra vestibulum nec, rutrum id leo. Nam ornare venenatis ex, non tempus felis malesuada in. Etiam congue suscipit sollicitudin. Proin aliquam elementum lorem, ac accumsan sem interdum eget. Nunc elit leo, tempus at elit id, pellentesque gravida augue.
      '''),
      )),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {
          await showCupertinoModalPopup(
            barrierDismissible: true,
            context: context,
            builder: (context) {
              return Container(
                height: 250,
                color: Colors.white,
                child: Column(
                  children: [
                    Flexible(
                      child: CupertinoDatePicker(
                        mode: CupertinoDatePickerMode.date,
                        initialDateTime: DateTime.now(),
                        onDateTimeChanged: (date) {
                          setState(() {
                            _date = date;
                          });
                        },
                      ),
                    ),
                  ],
                ),
              );
            },
          );
        },
      ),
    );
  }
}
```
</details>

**Comment by jonahwilliams** [comment]
I can reproduce this. All I need to do is open the cupertino picker and the bottom right corner will glitch slightly. This fixes itself after I interact with the picker

**Comment by jonahwilliams** [comment]
Its not a synchronization issue, this appears to be a partial repaint problem - at least for me anyway. I can fix this by adding:

```
<key>FLTDisablePartialRepaint</key>
<true />
```

To the Runner/Info.plist. FYI @knopp

**Comment by knopp** [comment]
@jonahwilliams, I don't seem to b able to reproduce this, at least not in simulator. When device are you testing this with? I assume on main? Also, just in case, could you try runing it with `<key>FLTEnableImpeller</key><false/>`. Partial repaint does work slightly differently with impeller and this might help narrowing it down.

**Comment by jonahwilliams** [comment]
If you run the example app, open the Cupertino picker, look in the bottom right corner.

**Comment by knopp** [comment]
@jonahwilliams, I was able to reproduce the issue. It seems that the whole picker renders wrongly with impeller, possibly out of calculated bounds too which would cause issues with partial repaint. Here's a GIF of impeller vs skia:

![partial-repaint](https://github.com/flutter/flutter/assets/96958/09aee638-f2b7-4f9e-bcca-0586c4336fa0)

**Comment by jonahwilliams** [comment]
Update: the depth changes did not land in 3.19 so this probably isn't https://github.com/flutter/flutter/issues/144333. Instead I'd guess something like https://github.com/flutter/engine/pull/45150

**Comment by bdero** [comment]
I think [this repro](https://github.com/flutter/flutter/issues/144331#issuecomment-1983284811) is https://github.com/flutter/flutter/issues/144333. The original post looks like a distinct issue.

**Comment by jonahwilliams** [comment]
Status: we're still unable to reproduce this issue. I landed a speculative revert of a related change. Without confirmation that this is or isn't helping we don't have enough information to proceed or put together a speculative cherry pick

**Comment by mtallenca** [comment]
I have a large app that I'm able to reproduce this bug on older iOS devices that are not able to run on iOS 17 (i.e. iPhone X, iPad mini 4). On newer devices I do not see this issue. Nor do I see the issue with small sample apps. My app is currently building with Flutter 3.19.5.  I'd love to test the speculative revert but there are numerous pub packages that we are using that need changes to work with current stable branch.  Would I be able to get a patched 3.19.5 engine to test with?

**Comment by jonahwilliams** [comment]
I could create a branch of 3.19.5 that has this fix applied, but I don't have the means to distribute binaries unless we do a release. Can you test with a local engine build @mtallenca ?

**Comment by mtallenca** [comment]
Haven't done a local engine build before - but I can do that.

**Comment by jonahwilliams** [comment]
The branch is here: https://github.com/flutter/engine/pull/51855 . You can either use git checkout or the gh tool to download that.

**Comment by mtallenca** [comment]
I can confirm the fix works. Tests done with release build on iOS, iPad mini 4

engine as is - app running on iPad mini 4 - https://share.icloud.com/photos/0707u9GCkVVuwSv58N7ehEzIg
engine#41855 - running on same iPad mini 4 - https://share.icloud.com/photos/0a918oBzOkf62y4QsYTGa0T-Q

**Comment by AliakseiT** [comment]
It is fixed in Flutter 3.19.6

**Comment by jonahwilliams** [comment]
We've cherry picked the fix for this into the last flutter stable. Closing.
